urbansearch/gathering/gathering.py

```diff
@@ -0,0 +1,157 @@
+from urllib.parse import quote
+import config
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class PageDownloader(object):
+
+    """
+    PageDownloader class. Creates object for a downloader with functions
+    to download pages for a certain url. Also contains functions to parse
+    the downloaded data to plain text.
+    """
+
+    def __init__(self):
+        self.cc_data_prefix = config.get('gathering', 'cc_data')
+        self.cc_index_url = config.get('gathering', 'cc_index')
+        self.indices = []
+
+    def download_indices(self, url, collection):
+        """
+        Download indices corresponding to url from Common Crawl collection.
+        Store indices in this PageDownloader object.
+
+        :param url: The url in string format
+        :param collection: Name of the collection, e.g CC-Main-2015-27-index
+        """
+        enc_url = quote(url, safe='')
+        try:
+            req_timeout = config.get('gathering', 'request_timeout')
+            response = requests.get(self.cc_index_url + collection +
+                                    '?url=' + enc_url +
+                                    '&output=json', timeout=req_timeout)
+            indices = [json.loads(x) for x in
+                       response.content.strip().decode('utf-8').split('\n')]
+            # TODO Benchmark to check forward scanning string for status is
+            # faster than removing after JSON (Speed improvement card)
+            self._clean_indices(indices)
+            self.indices += indices
+
+        except requests.exceptions.ReadTimeout:
+            logger.warning("URL index request timed out")
+
+    def download_warc_part(self, index):
+        """
+        Download the part of the warc file from common crawl servers
+        using the JSON index.
+
+        :param index: index in JSON format
+        :return: Uncompressed part of warc file if responsecode for index is
+        200, otherwise None
+        """
+        req_timeout = config.get('gathering', 'request_timeout')
+        start, length = int(index['offset']), int(index['length'])
+        end = start + length - 1
+        try:
+            response = requests.get(self.cc_data_prefix + index['filename'],
+                                    headers={'Range': 'bytes={}-{}'.format(start, end)}, 
+                                    timeout=req_timeout)
+        except requests.exceptions.ReadTimeout:
+            logger.warning("Timeout while downloading warc part")
+            return None
+
+        # Response is compressed gz data, uncompress this using gzip
+        compressed_gz = io.BytesIO(response.content)
+        with gzip.GzipFile(fileobj=compressed_gz) as gz_obj:
+            data = gz_obj.read()
+
+        return data
+
+    def _useful_responsecode(self, index):
+        # Check responsecode of index to determine if it's useful to download
+        # the part. HTTP 200 is useful, other than 200 will be discarded.
+        if index:
```

Comment:
Not sure why, we would still need to evaluate and disregard None in calling functions right?